platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/migration/ConfigurationMigrationManager.java

```diff
@@ -60,130 +62,246 @@
-     * @param mBeanServer                 object used to register this object as an MBean
-     * @param configurationMigratables    list of {@link ConfigurationMigratable} services. Needs
-     *                                    to be kept up-to-date by the client of this class.
-     * @param dataMigratables             list of {@link DataMigratable} services. Needs
-     *                                    to be kept up-to-date by the client of this class.
+     * @param mBeanServer object used to register this object as an MBean
+     * @param migratables list of {@link Migratable} services. Needs
+     *                    to be kept up-to-date by the client of this class.
+     * @param system      the system service
+     * @throws IOError if unable to load the distribution version information.
      */
-    public ConfigurationMigrationManager(
-            @NotNull ConfigurationAdminMigration configurationAdminMigration,
-            @NotNull MBeanServer mBeanServer,
-            @NotNull List<ConfigurationMigratable> configurationMigratables,
-            @NotNull List<DataMigratable> dataMigratables) {
-        notNull(configurationAdminMigration, "ConfigurationAdminMigration cannot be null");
-        notNull(mBeanServer, "MBeanServer cannot be null");
-        notNull(configurationMigratables,
-                "List of ConfigurationMigratable services cannot be null");
-        notNull(dataMigratables, "List of DataMigratable services cannot be null");
-
-        this.configurationAdminMigration = configurationAdminMigration;
+    public ConfigurationMigrationManager(MBeanServer mBeanServer, List<Migratable> migratables,
+            SystemService system) {
+        Validate.notNull(mBeanServer, "invalid null bean server");
+        Validate.notNull(migratables, "invalid null migratables");
+        Validate.notNull(system, "invalid null system service");
         this.mBeanServer = mBeanServer;
-        this.configurationMigratables = configurationMigratables;
-        this.dataMigratables = dataMigratables;
+        this.migratables = migratables;
+        this.system = system;
+        try {
+            this.productVersion =
+                    ConfigurationMigrationManager.getProductVersion(new FileInputStream(Paths.get(
+                            System.getProperty("ddf.home"),
+                            ConfigurationMigrationManager.PRODUCT_VERSION_FILENAME)
+                            .toFile()));
+        } catch (IOException e) {
+            LOGGER.warn("unable to load version information; ", e);
+            throw new IOError(e);
+        }
+    }
+
+    public static <T> BinaryOperator<T> throwingMerger() {
+        return (u, v) -> {
+            throw new IllegalStateException(String.format("Duplicate key %s", u));
+        };
+    }
+
+    private static String getProductVersion(InputStream is) throws IOException {
```

Comment:
✏️ You can simplify by replacing this method with...
```java
private static String getProductVersion(Path path) throws IOException {
    Validate.isTrue(Files.exists(path), "Missing path");
    return Files.lines(path)
            .findFirst()
            .map(String::trim)
            .filter(s -> !s.isEmpty())
            .orElseThrow(() -> new IOException("missing product version information"));
}
```
and then just passing the Path in directly in your ctor.

Reply:
I like it a lot. Thanks. Did not put the Validate as I wanted the IOException if the file doesn't exist.